packages/utils/telemetry-utils/src/batchedTelemetryHelper.ts

```diff
@@ -0,0 +1,73 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+type ICustomDataMap = Record<string, number>;
+
+/**
+ * Helper class that accumulates custom data and sends it to the logger every time a specified number (i.e., `threshold`) of logs is reached.
+ */
+export class BatchedTelemetryHelper {
+	disposed: boolean = false;
+
+	// Stores value of the custom data passed into the logger.
+	private readonly customDataMap: Map<string, number> = new Map<string, number>();
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	/**
+	 * @param eventBase -
+	 * Custom properties to include in the telemetry performance event when it is written.
+	 * @param logger -
+	 * The logger to use to write the telemetry performance event.
+	 * @param threshold -
+	 * The number of logs to accumulate before sending the data to the logger.
+	 */
+	public constructor(
+		private readonly eventBase: ITelemetryGenericEventExt,
+		private readonly logger: ITelemetryLoggerExt,
+		private readonly threshold: number,
+	) {}
+
+	/**
+	 * Accumulates the custom data and sends it to the logger every time the number of logs reaches the threshold by calling `sendData()`.
+	 *
+	 * @param customData -
+	 * A record storing the custom data to be logged.
+	 */
+	public log(customData: ICustomDataMap): void {
+		for (const key of Object.keys(customData)) {
+			this.customDataMap.set(key, (this.customDataMap.get(key) ?? 0) + customData[key]);
+		}
+
+		this.counter++;
+
+		if (this.counter >= this.threshold) {
+			this.sendData();
+		}
+	}
+
+	private sendData(): void {
+		const customData = Object.fromEntries(
+			[...this.customDataMap.entries()].map(([key, value]) => [key, value / this.counter]),
+		);
+
+		// TODO: Add `average` name to the custom data.
+		const telemetryEvent: ITelemetryPerformanceEventExt = {
+			...this.eventBase,
+			...customData,
+		};
+
+		this.logger.sendPerformanceEvent(telemetryEvent);
+		this.counter = 0;
```

Comment:
Do we want to flush the existing metrics after each time we send the data? If not that should be either mentioned in the comments or we can add an optional parameter to reset the data on send to this classes constructor.

Reply:
https://github.com/microsoft/FluidFramework/pull/21325#discussion_r1633503697

More details.